### PR TITLE
chore: Migrate CI & CD to GitHub Actions

### LINKED ISSUE #319: migrate build & release to GitHub Actions
We've recently migrated the .NET project to GitHub Actions for build and environment based deployments.  Given the recent need to honor Azure DevOps security recommendations, it would require some work to appropriately setup the ability to build from Forks; so it's a good time to make this change now anyway.  We'd seen the build not trigger in #318

### PR DESCRIPTION
**Pull request by LeeDongGeon1996**
### Description

 - Migrate CI & CD to GitHub Actions

Fixes #319

### Checklist

- [x] My code follows the existing style, code structure, and naming taxonomy
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my own code and included any verifying manual calculations
- [x] I have added or updated unit tests that prove my fix is effective or that my feature works, and achieves sufficient code coverage.  New and existing unit tests pass locally and in the build (below) with my changes
- [x] My changes generate no new warnings and running code analysis does not produce any issues
- [x] I have added or run the performance tests that depict optimal execution times
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by LeeDongGeon1996** [comment]
GPT did it all...👍

**Comment by LeeDongGeon1996** [comment]
@DaveSkender  Can you check Azure Pipelines webhook config?
And are we going to keep using gitversion for deploying? or we can just tag and use tag as a package version.

**Comment by DaveSkender** [comment]
I've added the `Trusted Publisher Management` for two new environments `pypi` and `pypi-test`.  The former requires my review; whereas the later _should be_ a bit more open for action.

image
image

**Comment by DaveSkender** [comment]
Also, check for an equivalent auto-draft for the release notes, same or similar to:

https://github.com/DaveSkender/Stock.Indicators/blob/6c70e30eff7e9b8bbd0e37a3fbf006e585346b34/.github/workflows/deploy-package.yml#L157-L167

Note: this may need to be configured slightly different if the timing of tagging is different.  I didn't think through the scenario, so it may be okay with same implementation.

**Comment by DaveSkender** [comment]
I've also removed the required check for the Azure Pipelines build, but will need to add the new CI.yml variant to the branch protection rules once everything's working well here, and probably after this is merged.

**Comment by LeeDongGeon1996** [comment]
@DaveSkender Thanks for your work :)

>  The former requires my review; whereas the later should be a bit more open for action.

Yeah, that's why we need to set [deployment protection rules](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#deployment-protection-rules) or add [review step](https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments) before deployment

File: .github/workflows/deploy.yml
Comment (LeeDongGeon1996) [inline .github/workflows/deploy.yml:63]: I've added this configuration as you recommended. Please set PYPI_DOWNLOAD_PREFIX to `https://pypi.org/project/stock-indicators/`

Comment (DaveSkender) [reply .github/workflows/deploy.yml:63]: I've added this to the `pypi` environment, and also an equivalent one for the `pypi-test` URL, but suspect that'd never get used; just added it for parity.

image

Comment (DaveSkender) [reply .github/workflows/deploy.yml:63]: I also updated to the current `deploy.yml` filename on PyPI settings.

image

File: .github/workflows/deploy.yml
Comment (DaveSkender) [inline .github/workflows/deploy.yml:61]: This is `stock-indicators` on PyPI.

**Comment by DaveSkender** [comment]
I really only have one question, how is the release tag introduced on `main`?

**Review by DaveSkender — APPROVED** [review]

File: .github/workflows/deploy.yml
Comment (DaveSkender) [inline .github/workflows/deploy.yml:69]: Maybe I was wrong here.  I think you can create a Tag when you _manually_ create a [draft] Release record, then wait for it to deploy, then make the Release a non-Draft?  The only problem with that approach is that it wouldn't auto-generate the release notes.

Comment (DaveSkender) [reply .github/workflows/deploy.yml:69]: (other than bypassing branch protection on `main` and pushing a Tag directly)

Comment (DaveSkender) [reply .github/workflows/deploy.yml:69]: Oh, wait, you can manually add [automatically generated release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes)

Comment (DaveSkender) [reply .github/workflows/deploy.yml:69]: > I really only have one question, how is the release tag introduced on main?

To answer my own question, you can manually create a Draft Release, add a new custom Tag with the version number, add auto-generated release notes; then Save; wait for deployment, remove Draft status from Release note.  Is that the approach here, or did you have something else in mind?

Comment (LeeDongGeon1996) [reply .github/workflows/deploy.yml:69]: Ah, I was planning to push tag from my local repository.

Comment (DaveSkender) [reply .github/workflows/deploy.yml:69]: Feature branch tags get squashed (removed) when you merge to `main`, in our normal merge strategy.  We'd have to push a solitary tag directly to `main` if we want to keep Squash merge limitations (which is nicer commit history).

**Review by DaveSkender — COMMENTED** [review]
Also, we'd need to pull out the unused `SemVer` stuff like `gitversion.yml`, how its written in `contributing.md`, and any other integrations that might exist in `setup.py` or other places where the version number is stamped.

File: .github/workflows/deploy.yml
Comment (LeeDongGeon1996) [reply .github/workflows/deploy.yml:69]: You mean I cannot push a tag directly, because I'm working on forked repository. hmm okay.
Let's just remove `Tag and draft release note` step and use GH native feature!

Comment (DaveSkender) [reply .github/workflows/deploy.yml:69]: Go for it, let's give is a try.  Just a couple other removals for cleanup below too.

**Comment by DaveSkender** [comment]
How does the `test.pypi.org` publish get it's version number?

**Comment by LeeDongGeon1996** [comment]
You're right! I'll draft fixed `contributing.md` and then you correct that please.

**Comment by LeeDongGeon1996** [comment]
> How does the test.pypi.org publish get its version number?

`setuptools_scm` does everything. It recognize the latest tag on repository. (AFAIK, it uses `git describe`)

**Comment by LeeDongGeon1996** [comment]
Yeah as I know, setuptool_scm resolve that problem by creating unique version number, let me double check.

**Comment by DaveSkender** [comment]
Let me know when you're done with `contributing.md` and I'll take an editorial pass on it.

**Comment by LeeDongGeon1996** [comment]
Yeah, I've already done!

**Review by DaveSkender — APPROVED** [review]
All good on my end.  Merge when ready.

**Pull request merged**
